File: java/src/Basket/Basket.java
package Basket;

import java.io.PrintStream;
import java.util.HashMap;
import java.util.Map;

import Products.Catalogue;
import Products.Sku;
import Warehouse.Warehouse;

public class Basket {

	private Map<String, BasketItem> items;
	
	public Basket() {
		items = new HashMap<String, BasketItem>();
	}

	public void list(PrintStream out, Catalogue catalogue) {
		if (items.isEmpty()) {
			System.out.println("Your basket is empty");
			return;
		}
		for (String skuId : items.keySet()) {						// TODO -- sort by SKU
			Sku sku = catalogue.lookup(skuId);
			Integer count = items.get(skuId).count;						// TODO -- print item titles
			out.printf("%dp x%d\t%s\n", sku.price, count, sku.title);
		}															// TODO -- print total
	}

	public void add(String skuId, Catalogue catalogue, int numItems) {
		Sku sku = catalogue.lookup(skuId);
		if (items.containsKey(skuId)) {
			BasketItem existing = items.get(skuId);
			items.put(skuId, new BasketItem(skuId, existing.title, existing.price, existing.count + numItems));
		} else {
			items.put(skuId, new BasketItem(skuId, sku.title, sku.price, numItems));
		}
	}

	public void checkout(Warehouse warehouse) {
		if (items.isEmpty()) {
			System.err.println("Your basket is empty!");
			return;
		}
		for (String sku : items.keySet())
			warehouse.fulfill(sku, items.get(sku).count);
		items = new HashMap<String, BasketItem>();
		System.out.println("All items checked out.");			// TODO -- print total
	}

}

File: ruby/lib/basket/basket.rb
require_relative './basket_item'

module Basket

	class Basket

		def initialize
			@items = {}
		end

		def list
			sorted_skus.map {|sku| @items[sku] }
		end

		def add(sku_id, catalogue, numItems)
			sku = catalogue.lookup(sku_id)
			if @items.include?(sku_id)
				existing = @items[sku_id]
				@items[sku_id] = BasketItem.new(sku_id, existing.title, existing.price, existing.count + numItems)
			else
				@items[sku_id] = BasketItem.new(sku_id, sku.title, sku.price, numItems)
			end
		end

		def checkout(warehouse)
			if @items.empty?
				$stderr.puts "Your basket is empty!"
				return
			end
			@items.keys.each do |sku|
				warehouse.fulfill(sku, @items[sku].count)
			end
			@items = {}
			printf("All items checked out. Total price £%5.02f\n", current_total / 100.0)
		end

		private

		def sorted_skus
			@items.keys.sort
		end

		def current_total
			total = 0
			@items.values.each do |item|
				total += item.count * item.price
			end
			if total > 2000
				total -= (total/10)
			end
			total
		end

	end
end
